core/src/main/java/jenkins/model/logging/Loggable.java

```diff
@@ -0,0 +1,85 @@
+package jenkins.model.logging;
+
+import jenkins.model.logging.impl.FileLogStorage;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.Beta;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.io.File;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Interface which indicates that custom logging is applicable to the object.
+ * @author Oleg Nenashev
+ * @since TODO
+ * @see LogBrowser
+ * @see LoggingMethod
+ */
+@Restricted(Beta.class)
+public interface Loggable {
+
+    @Nonnull
+    default LoggingMethod getLoggingMethod() {
+        return LoggingMethodLocator.locate(this);
+    }
+
+    /**
+     * Determines a default logger to be used.
+     * @return Default logger.
+     */
+    @Nonnull
+    LoggingMethod getDefaultLoggingMethod();
+
+    @Nonnull
+    default LogBrowser getLogBrowser() {
+        return LoggingMethodLocator.locateBrowser(this);
+    }
+
+    /**
+     * Determines a default log browser to be used.
+     * @return Default log browser.
+     */
+    @Nonnull
+    LogBrowser getDefaultLogBrowser();
+
+    /**
+     * Returns {@code true} if the log file is no longer being updated.
+     */
+    public boolean isLoggingFinished();
+
+    /**
+     * Returns charset to be used.
+     * New implementations are recommended to use {@code UTF-8}-only (default),
+     * but the method can be overridden by legacy implementations.
```

Comment:
Cc @casz who also wanted to get rid of non-UTF-8 logging iirc

Reply:
I concur with jesse.

Reply:
Current design: https://github.com/jenkinsci/jep/tree/master/jep/207#encoding-of-log-files

```


Currently Jenkins does not set limitations for encoding while doing logging. Any charsets may be used on agent and master sides, and it is hard to manage them.

Although it is expected that all logs eventually switch to UTF-8 (see the JEP-206 proposal for Pipeline), in meantime external logging may be performed in different encodings.

*  Loggable implementations can define the charset to be used
* Logging Method and Logging Browser implementations may implement support of charsets or reject them, it is up to the implementation
* If the implementation does not support a charset, `ExternalLoggingMethodLocator` can skip the logging method
```

I believe this is a more conservative approach. I do not want to fix Freestyle logging in this change, because it may introduce serious regressions.

If you prefer, I could add a `supportsNonUTF8()` method (defaults to `false`) so that `LoggingMethodLocator` checks support automatically.

Reply:
> I do not want to fix Freestyle logging in this change, because it may introduce serious regressions.

Well, the same could be said for JEP-206, but my feeling was that it was worth risking regressions in the interest of making the system design simpler and clearer in the long run.

Reply:
(see #3231)